test/browser/component.js

```diff
@@ -843,3 +844,284 @@ describe(`Component with required attrs`, function () {
+    });
+  });
+
+  describe(`lifecycle`, function () {
+    beforeEach(async function () {
+      document.body.innerHTML = ``;
+      await nextAnimationFrame();
+    });
+
+    it(`fails to connect when a context declared in config does not have a default context by itself or from any context ancestor`, async function () {
+      // modern browsers with native Custom Elements support will emit a global error event
+      const errors = [];
+      window.uncaughtErrorFilter = (errorEvent) => {
+        errors.push(errorEvent.message);
+        return true;
+      };
+
+      try {
+        const widget = document.createElement(`themed-widget`);
+        document.body.appendChild(widget);
+      } catch (err) {
+        // older browsers that rely on a Custom Elements polyfill will throw an error directly
+        errors.push(err.message);
+      }
+      await nextAnimationFrame();
+
+      expect(errors).to.have.lengthOf(1);
+      expect(errors[0]).to.contain(`A "theme" context is not available`);
+
+      delete window.uncaughtErrorFilter;
+    });
+
+    it(`executes bindToComponent callback when connected`, async function () {
+      const widget1 = document.createElement(`slotted-load-counter-widget`);
+      const counter = widget1.getConfig(`defaultContexts`).loadCounter;
+      const bindToComponentSpy = sinon.spy(counter, `bindToComponent`);
+      document.body.appendChild(widget1);
+      await nextAnimationFrame();
+      expect(counter.getCount()).to.equal(1);
+      expect(bindToComponentSpy.getCall(0).args[0]).to.equal(widget1);
+
+      const widget2 = document.createElement(`slotted-load-counter-widget`);
+      widget1.appendChild(widget2);
+      await nextAnimationFrame();
+      expect(counter.getCount()).to.equal(2);
+      expect(bindToComponentSpy.getCall(1).args[0]).to.equal(widget2);
+
+      const widget3 = document.createElement(`slotted-load-counter-widget`);
+      widget2.appendChild(widget3);
+      await nextAnimationFrame();
+      expect(counter.getCount()).to.equal(3);
+      expect(bindToComponentSpy.getCall(2).args[0]).to.equal(widget3);
+
+      bindToComponentSpy.restore();
+    });
+
+    it(`executes bindToComponent callback each time the same component is repeatedly connected`, async function () {
+      const app = document.createElement(`slotted-load-counter-widget`);
+      const counter = app.getConfig(`defaultContexts`).loadCounter;
+      const bindToComponentSpy = sinon.spy(counter, `bindToComponent`);
+      document.body.appendChild(app);
+      await nextAnimationFrame();
+      expect(counter.getCount()).to.equal(1);
+      expect(bindToComponentSpy.getCall(0).args[0]).to.equal(app);
+
+      const widget = document.createElement(`slotted-load-counter-widget`);
+      app.appendChild(widget);
+      await nextAnimationFrame();
+      expect(counter.getCount()).to.equal(2);
+      expect(bindToComponentSpy.getCall(1).args[0]).to.equal(widget);
```

Comment:
might be useful to precede each of these checks with an explicit verification of the number of calls
```suggestion
      expect(bindToComponentSpy).to.have.callCount(2);
      expect(bindToComponentSpy.getCall(1).args[0]).to.equal(widget);
```
(I think we have sinon-chai here, otherwise the syntax will be a little different)